### DISCUSSION TITLE
Partial resource update using HTTP PATCH

### DISCUSSION THREAD

**Discussion by jeromepin** [fastapi/fastapi#8258] [Questions] [answered]
**Description**

How can I do a partial update of a given resource ? A PATCH is always blocked because of pydantic data validation, as the body doesn't match the schema defined (obviously !).

Is there something I'm missing ? Because I would like to not retrieve the whole object in order to do a simple update on one given field. Either by specifying a partial object or a json patch ?

**Comment by tiangolo** [comment, marked as answer]
You can have an additional Pydantic model with all the fields set as optional (with `= None`).

And then, only update the fields that are passed.

For example, check the project generator and how the update is handled: https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/api/api_v1/endpoints/user.py#L138

And the model has all the fields optional: https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/models/user.py#L25-L26

Doing it with a `PATCH` would be the most semantically correct, but FastAPI doesn't really enforce which methods you use or how.

In fact, the project generators tend to use `PUT` for updates even though they act more like a `PATCH` (a partial update instead of a full resource replacement). Just because many frontend and other devs aren't very familiar with `PATCH`.

But the trick in this case is actually in the Pydantic models and what you define as required or not.

**Comment by jeromepin** [comment]
That's what I had in mind too, but I thought it was a workaround. I'm going to do things that way then. Thank you !

**Comment by tiangolo** [comment]
Cool! Thanks for coming back and closing it.

**Comment by Dock1100** [comment]
Hi, with `patch` request we can omit some fields (not sending it at all, so that field will not be updated), and also send `"field": null` to set field's value to `None`. How in this case I can distinguish if field really sent as null, or it was omitted and has default value `None` set by pydantic model?

**Comment by tiangolo** [comment]
@Dock1100 please check the docs: https://fastapi.tiangolo.com/tutorial/body-updates/#partial-updates-with-patch

**Comment by Dock1100** [comment]
@tiangolo my bad. `exclude_unset=True` is what I was looking for, thanks.

**Comment by coder-amogh** [comment]
Hi, I'm having issues with the partial updates...

These are my files:

### models.py

```
class User(Base):
	__tablename__ = "users"

	id = Column(String(64), primary_key = True, index = True, nullable = False)
	email = Column(String(64), unique = True, index = True, nullable = False)
	password = Column(String(64), nullable = False)
	is_active = Column(Boolean, server_default = "1", nullable = False) # https://stackoverflow.com/questions/12045698/sqlalchemy-boolean-value-is-none#comment69750602_12045897
	added_at = Column(DateTime(), server_default = func.now(), nullable = False)

	items = relationship("Item", back_populates = "user")

class Item(Base):
	__tablename__ = "items"

	id = Column(String(64), primary_key = True, index = True, nullable = False)
	name = Column(String(128), index = True, nullable = False)
	description = Column(String(1024), index = True, nullable = True)
	color = Column(Enum(ItemColorEnum), nullable = True)
	user_id = Column(String(64), ForeignKey("users.id"), nullable = False)
	added_at = Column(DateTime(), server_default = func.now(), nullable = False)

	user = relationship("User", back_populates = "items")
```

### schemas.py

```
class ItemColorEnum(str, enum.Enum):
	red = "red"
	green = "green"
	blue = "blue"

class ItemBase(BaseModel):
	name: Optional[str]
	description: Optional[str] = Field(None, example = "Buy 5 pens till EOD.")
	color: Optional[ItemColorEnum] = Field(None, example = "blue")

class ItemCreate(ItemBase):
	name: str = Field(None, example = "Buy 5 pens till EOD.")

class ItemUpdate(ItemBase):
	pass

```

### main.py

```
@app.patch("/update-item/{item_id}", status_code = 200)
def update_item(item_data: schemas.ItemUpdate, item_id: str, db: Session = Depends(get_db), user_id = Depends(auth_handler.auth_wrapper)):
	db_item = crud.get_item(db, item_id)

	if db_item:
		if db_item.user_id != user_id:
			raise HTTPException(status_code = 403, detail = "You are forbidden to make this request!")		
	else:
		raise HTTPException(status_code = 404, detail = "Item not found!")

	update_item_result = crud.update_item(db, item_id, item_data, db_item)

	if not update_item_result:
		raise HTTPException(status_code = 500, detail = "Server error...")

	return {"updated": True}

```

### crud.py

```

def get_item(db, item_id: str):
	item = db.query(models.Item).filter(models.Item.id == item_id).first()
	return item

def update_item(db, item_id, item_data: schemas.ItemUpdate, db_item):
	stored_item_object_schema = schemas.ItemUpdate(name = db_item.name, description = db_item.description, color = db_item.color)

	updated_item_dict = item_data.dict(exclude_unset = True)

	updated_item_object_schema = stored_item_object_schema.copy(update = updated_item_dict)

	query = db.query(models.Item).filter(models.Item.id == item_id).update(updated_item_object_schema)

	db.commit()

	return True
```

I have problems with the update item function. Actually, I'm unsure how to proceed. The goal is to allow the user to only send the fields that need to be updated. As you can see from the `models.py` file, the title/name for the item is compulsory and shouldn't be null. However, if I run the `main.py` and then send `{"name": null}` to try to remove the item name from the database, it returns `Internal Server Error`.

What's good is that it doesn't allow the update in the database, but what's bad is it doesn't return a clean JSON response. Is it possible to somehow catch this error and return a JSON response?

P.S: Can you (or anyone for that matter) suggest any improvements in my code? It's my first time with FastAPI, but I'm definitely going to use it for all my purposes. It's just a great job!

**Comment by tiangolo** [comment]
@coder-amogh if you still have problems, please create a new issue following the template. 🤓 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
